presto-main/src/main/java/com/facebook/presto/operator/GroupedTopNBuilder.java

```diff
@@ -0,0 +1,654 @@
+                groupedRows.set(groupId, rows);
+            }
+            else {
+                // update an existing group;
+                // remove the memory usage for this group for now; add it back after update
+                memorySizeInBytes -= rows.getEstimatedSizeInBytes();
+            }
+
+            if (rows.size() < topN) {
+                // still have space for the current group
+                Row row = new Row(newPageId, position);
+                rows.enqueue(row);
+                newPageReference.reference(row);
+            }
+            else {
+                // may compare with the topN-th element with in the heap to decide if update is necessary
+                Row previousRow = rows.first();
+                Row newRow = new Row(newPageId, position);
+                if (comparator.compare(newRow, previousRow) < 0) {
+                    // update reference and the heap
+                    rows.dequeue();
+                    PageReference previousPageReference = pageReferences.get(previousRow.getPageId());
+                    previousPageReference.dereference(previousRow.getPosition());
+                    newPageReference.reference(newRow);
+                    rows.enqueue(newRow);
+
+                    // compact a page if it is not the current input page and the reference count is below the threshold
+                    if (previousPageReference.getPage() != newPage &&
+                            previousPageReference.getUsedPositionCount() * COMPACT_THRESHOLD < previousPageReference.getPage().getPositionCount() &&
+                            !pagesToCompact.contains(previousRow.getPageId())) {
+                        pagesToCompact.add(previousRow.getPageId());
+                    }
+                }
+            }
+
+            memorySizeInBytes += rows.getEstimatedSizeInBytes();
+        }
+
+        // may compact the new page as well
+        if (newPageReference.getUsedPositionCount() * COMPACT_THRESHOLD < newPage.getPositionCount()) {
+            verify(!pagesToCompact.contains(newPageId));
+            pagesToCompact.add(newPageId);
+        }
+
+        compactPages(pagesToCompact.iterator());
+    }
+
+    private void compactPages(Iterator<Integer> pageIds)
+    {
+        while (pageIds.hasNext()) {
+            int pageId = pageIds.next();
+            PageReference pageReference = pageReferences.get(pageId);
+            if (pageReference.getUsedPositionCount() == 0) {
+                pageReferences.set(pageId, null);
+                emptyPageReferenceSlots.enqueue(pageId);
+                memorySizeInBytes -= pageReference.getEstimatedSizeInBytes();
+            }
+            else {
+                memorySizeInBytes -= pageReference.getEstimatedSizeInBytes();
+                pageReference.compact();
+                memorySizeInBytes += pageReference.getEstimatedSizeInBytes();
+            }
+        }
+    }
+
+    private class Row
```

Comment:
Add a comment that this is a pointer to a row in a page.  The actual position in the page is mutable because as pages are compacted, the position will change.